File: captchax/sdk/php/tests/CaptchaXClientTest.php
<?php

declare(strict_types=1);

namespace CaptchaX\SDK\Tests;

use CaptchaX\SDK\CaptchaConfig;
use CaptchaX\SDK\CaptchaXClient;
use CaptchaX\SDK\CaptchaXException;
use CaptchaX\SDK\ApiVersion;
use CaptchaX\SDK\CharPosition;
use PHPUnit\Framework\TestCase;

class CaptchaXClientTest extends TestCase
{
    public function testCreateClientWithConfig(): void
    {
        $config = new CaptchaConfig('https://captchax.example.com', 'test-app');
        $client = new CaptchaXClient($config);
        $this->assertSame(ApiVersion::V1, $client->getApiVersion());
    }

    public function testCreateClientWithStaticFactory(): void
    {
        $client = CaptchaXClient::create('https://captchax.example.com');
        $this->assertInstanceOf(CaptchaXClient::class, $client);
    }

    public function testSetAppId(): void
    {
        $client = CaptchaXClient::create('https://captchax.example.com');
        $client->setAppId('new-app-id');
        $this->assertTrue(true);
    }

    public function testSetApiVersion(): void
    {
        $client = CaptchaXClient::create('https://captchax.example.com');
        $this->assertSame(ApiVersion::V1, $client->getApiVersion());

        $client->setApiVersion(ApiVersion::V2);
        $this->assertSame(ApiVersion::V2, $client->getApiVersion());
    }

    public function testCreateClientInfo(): void
    {
        $client = CaptchaXClient::create('https://captchax.example.com');
        $info = $client->createClientInfo(['custom' => 'value']);

        $this->assertNotEmpty($info);
        $data = json_decode($info, true);
        $this->assertArrayHasKey('platform', $data);
        $this->assertArrayHasKey('timestamp', $data);
        $this->assertSame('value', $data['custom']);
    }
}

class CaptchaConfigTest extends TestCase
{
    public function testDefaultValues(): void
    {
        $config = new CaptchaConfig('https://captchax.example.com');
        $this->assertSame('https://captchax.example.com', $config->getBaseUrl());
        $this->assertNull($config->getAppId());
        $this->assertSame(10000, $config->getTimeout());
        $this->assertSame(3, $config->getRetryTimes());
        $this->assertSame(ApiVersion::V1, $config->getApiVersion());
    }

    public function testWithMethods(): void
    {
        $config = (new CaptchaConfig('https://captchax.example.com'))
            ->withAppId('test-app')
            ->withTimeout(5000)
            ->withRetryTimes(2)
            ->withApiVersion(ApiVersion::V2);

        $this->assertSame('test-app', $config->getAppId());
        $this->assertSame(5000, $config->getTimeout());
        $this->assertSame(2, $config->getRetryTimes());
        $this->assertSame(ApiVersion::V2, $config->getApiVersion());
    }
}

class CaptchaXExceptionTest extends TestCase
{
    public function testExceptionWithDefaults(): void
    {
        $exception = new CaptchaXException('Test error');
        $this->assertSame('Test error', $exception->getMessage());
        $this->assertSame(500, $exception->getErrorCode());
        $this->assertSame(500, $exception->getStatusCode());
    }

    public function testExceptionWithCustomValues(): void
    {
        $exception = new CaptchaXException('Test error', 400, 400, ['field' => 'value']);
        $this->assertSame(400, $exception->getErrorCode());
        $this->assertSame(400, $exception->getStatusCode());
        $this->assertSame(['field' => 'value'], $exception->getDetails());
    }
}

class CharPositionTest extends TestCase
{
    public function testCharPositionCreation(): void
    {
        $pos = new CharPosition('A', 100, 50);
        $this->assertSame('A', $pos->getChar());
        $this->assertSame(100, $pos->getX());
        $this->assertSame(50, $pos->getY());
    }

    public function testCharPositionToArray(): void
    {
        $pos = new CharPosition('B', 200, 75);
        $arr = $pos->toArray();

        $this->assertSame([
            'char' => 'B',
            'x' => 200,
            'y' => 75,
        ], $arr);
    }
}

File: captchax/sdk/java/src/test/java/com/captchax/sdk/CaptchaXClientTest.java
package com.captchax.sdk;

import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.*;

public class CaptchaXClientTest {

    @Test(expected = IllegalArgumentException.class)
    public void testConstructorWithNullBaseUrl() {
        new CaptchaXClient((String) null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructorWithEmptyBaseUrl() {
        new CaptchaXClient("");
    }

    @Test
    public void testConstructorWithValidConfig() {
        CaptchaConfig config = new CaptchaConfig.Builder()
                .baseUrl("https://captchax.example.com")
                .appId("test-app")
                .timeout(5000)
                .retryTimes(2)
                .apiVersion(ApiVersion.V2)
                .build();

        CaptchaXClient client = new CaptchaXClient(config);
        assertEquals(ApiVersion.V2, client.getApiVersion());
    }

    @Test
    public void testSetAppId() {
        CaptchaXClient client = new CaptchaXClient("https://captchax.example.com");
        client.setAppId("new-app-id");
    }

    @Test
    public void testSetApiVersion() {
        CaptchaXClient client = new CaptchaXClient("https://captchax.example.com");
        assertEquals(ApiVersion.V1, client.getApiVersion());

        client.setApiVersion(ApiVersion.V2);
        assertEquals(ApiVersion.V2, client.getApiVersion());
    }

    @Test
    public void testCreateClientInfo() {
        CaptchaXClient client = new CaptchaXClient("https://captchax.example.com");
        String info = client.createClientInfo(null);
        assertNotNull(info);
        assertTrue(info.contains("platform"));
        assertTrue(info.contains("timestamp"));

        Map<String, Object> extra = new HashMap<>();
        extra.put("customField", "value");
        info = client.createClientInfo(extra);
        assertTrue(info.contains("customField"));
        assertTrue(info.contains("value"));
    }
}

class CaptchaXExceptionTest {

    @Test
    public void testExceptionWithMessage() {
        CaptchaXException exception = new CaptchaXException("Test error");
        assertEquals("Test error", exception.getMessage());
        assertEquals(500, exception.getCode());
        assertEquals(500, exception.getStatusCode());
    }

    @Test
    public void testExceptionWithCode() {
        CaptchaXException exception = new CaptchaXException("Test error", 400, 400);
        assertEquals("Test error", exception.getMessage());
        assertEquals(400, exception.getCode());
        assertEquals(400, exception.getStatusCode());
    }

    @Test
    public void testExceptionWithDetails() {
        CaptchaXException exception = new CaptchaXException("Test error", 400, 400, "details");
        assertEquals("details", exception.getDetails());
    }

    @Test
    public void testExceptionWithCause() {
        Exception cause = new RuntimeException("cause");
        CaptchaXException exception = new CaptchaXException("Test error", cause);
        assertEquals(cause, exception.getCause());
    }
}

class ApiVersionTest {

    @Test
    public void testApiVersionValues() {
        assertEquals("v1", ApiVersion.V1.getValue());
        assertEquals("v2", ApiVersion.V2.getValue());
    }
}

class CaptchaConfigTest {

    @Test
    public void testConfigBuilder() {
        CaptchaConfig config = new CaptchaConfig.Builder()
                .baseUrl("https://captchax.example.com")
                .appId("test-app")
                .timeout(5000)
                .retryTimes(2)
                .apiVersion(ApiVersion.V2)
                .build();

        assertEquals("https://captchax.example.com", config.getBaseUrl());
        assertEquals("test-app", config.getAppId());
        assertEquals(5000, config.getTimeout());
        assertEquals(2, config.getRetryTimes());
        assertEquals(ApiVersion.V2, config.getApiVersion());
    }

    @Test
    public void testConfigDefaults() {
        CaptchaConfig config = new CaptchaConfig("https://example.com");
        assertEquals(10000, config.getTimeout());
        assertEquals(3, config.getRetryTimes());
        assertEquals(ApiVersion.V1, config.getApiVersion());
    }
}
